include/oneapi/dpl/internal/histogram_impl.h

```diff
@@ -0,0 +1,117 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_HISTOGRAM_IMPL_H
+#define _ONEDPL_HISTOGRAM_IMPL_H
+
+#include "function.h"
+#include "histogram_extension_defs.h"
+#include "histogram_binhash_utils.h"
+#include "../pstl/iterator_impl.h"
+
+#if _ONEDPL_HETERO_BACKEND
+#    include "oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h"
+#endif
+
+namespace oneapi
+{
+namespace dpl
+{
+
+namespace __internal
+{
+
+//Only supported for dpcpp backend
+#if _ONEDPL_BACKEND_SYCL
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename _Size,
+          typename _IdxHashFunc, typename... _Range>
+inline oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<_ExecutionPolicy>::type>
+__pattern_histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _RandomAccessIterator2 __histogram_first, const _Size& __num_bins, _IdxHashFunc __func,
+                    _Range&&... __opt_range)
+{
+    //If there are no histogram bins there is nothing to do
+    if (__num_bins > 0)
+    {
+        oneapi::dpl::__par_backend_hetero::__parallel_histogram(::std::forward<_ExecutionPolicy>(exec), __first, __last,
+                                                                __histogram_first, __num_bins, __func,
+                                                                ::std::forward<_Range>(__opt_range)...);
+    }
+}
+#endif // _ONEDPL_BACKEND_SYCL
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename _Size,
+          typename _IdxHashFunc, typename... _Range>
+inline oneapi::dpl::__internal::__enable_if_host_execution_policy<typename ::std::decay<_ExecutionPolicy>::type>
+__pattern_histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _RandomAccessIterator2 __histogram_first, const _Size& __num_bins, _IdxHashFunc __func,
+                    _Range&&... __opt_range)
+{
+    assert(false && "Histogram API is not currently supported for host parallel policies");
```

Comment:
This might work better as a `static_assert`, since any calls to this overload with a host execution policy is known at compile time to be an error.

Reply:
Aha, thanks for pointing that out.  I'll try it.  I agree a static assert is far superior if we can get it to work.

Reply:
Yes, this seems to have fixed it (although the CI seems to be having other problems unrelated to this PR now...)